racket/collects/racket/bytes.rkt

```diff
@@ -2,9 +2,22 @@
 
 (provide bytes-append* bytes-join)
 
+(define (check-bytes-list* l name)
+  (let loop ((l l) (p null))
+    (cond ((null? (cdr l))
+           (define tail (car l))
+           (or (and (list? tail) (andmap bytes? tail))
+               (raise-argument-error name "(listof bytes?)" tail))
+           (foldl cons tail p))
+          (else
+           (define e (car l))
+           (or (bytes? e)
+               (raise-argument-error name "bytes?" e))
+           (loop (cdr l) (cons e p))))))
+
```

Comment:
A direct `cons` on the recursion seems simpler here — and it should be faster, although that's unlikely to matter:

```
(define (check-bytes-list* l name)
  (let loop ([l l])
    (cond
      [(null? (cdr l))
       (define tail (car l))
       (unless (and (list? tail) (andmap bytes? tail))
         (raise-argument-error name "(listof bytes?)" tail))
       tail]
      [else
       (define e (car l))
       (unless (bytes? e)
         (raise-argument-error name "bytes?" e))
       (cons e (loop (cdr l)))])))
```